### ISSUE TITLE
Persisting additional metadata somewhere?

### ISSUE THREAD

**Issue by tusharmath** [hashicorp/terraform#17430]
First of all terraform has made my life awfully simple, thanks a ton for making this amazing tool!

I am trying to implement `BLUE` `GREEN` deployments using terraform. I have two fleets — `blue` and `green` and I make a deployment on them alternatively while also scaling-in and scaling-out using terraform. Everything is perfect except that I need to remember which fleet was deployed last and make sure the new deployment happens on the other fleet. 

Is there a way to do using some logic or maybe by maintaining some external state?

**Comment by apparentlymart** [comment]
Hi @tusharmath!

Terraform already has a number of options for writing arbitrary information into several places, such as metadata tags on the resources themselves (where supported by the remote system), values in Consul, etc.

For blue/green scenarios I have also seen this done entirely within Terraform config by just checking in the cluster selection as part of the config into version control.

Without more details it's hard to know what sort of thing you need here. If you can share a real example we can see if there's already a solution available or see if we could add something to Terraform to support your use-case better.

Thanks!

**Comment by jbardin** [comment]
Hi @tusharmath,

If this is for manual verification, a simple option is to store the deployed value to an output in the root module. Possibly something like:

```
output "deployment_color" {
  value = "${var.color}
}
```

**Comment by apparentlymart** [comment]
A different pattern I've seen before (and used myself in a previous job) is to place the cluster config in a module and instantiate it multiple times:

```hcl
module "cluster-20180219" {
  source = "(cluster module address)"
  ami = "ami-abc123"
  instances = 3
}

module "cluster-20180219" {
  source = "(cluster module address)"
  ami = "ami-abc567"
  instances = 5
}
```

Then each time you want to roll a change you add a second module block, apply to create the new cluster, verify the release, remove the old module, and roll again.

An advantage of this approach is that you can also switch to a new version of the module itself by changing the source on the new block. The roll process can be use both for AMI changes and for config changes, considering each module block as immutable once it is added, until it is removed.

With Terraform as it currently stands I would either just manually edit the configs like above (and indeed did exactly that, albeit for a config that changed only once every couple months or so) or wrap some automation to generate module configurations programmatically based on some external input... such automation could then in principle step through the create/verify/update steps by scripting Terraform runs, especially if there is some automatic way to check if the new cluster is working.

My initial reaction is that this is something that ought to be built _around_ Terraform rather than within Terraform itself, but I will think a little about what Terraform could do itself here to help with this, in principle. If you have some thoughts yourself I'd love to hear them!

**Comment by tusharmath** [comment]
>My initial reaction is that this is something that ought to be built around Terraform rather than within Terraform

I agree completely here.

Giving a special resource where I can save random meta data and query it when needed with some default values if the resource isn't available, would make this thing possible and also solve a lot of other cases.

**Comment by tusharmath** [comment]
I feel I can totally use the [null_resource](null_resource) for persisting the values. I just need a way to read it also.

[null_resource]: https://www.terraform.io/docs/providers/null/resource.html

**Comment by copumpkin** [comment]
@apparentlymart when you say this:

> Terraform already has a number of options for writing arbitrary information into several places, such as metadata tags on the resources themselves (where supported by the remote system)

Is that generic across all resource types? I'm looking for roughly the equivalent of [CloudFormation's arbitrary structured metadata](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-metadata.html) that's preserved in the stack's state (so in the .tfstate file here), regardless of the underlying resource capabilities. Is that what you're talking about? I haven't been able to find much of that flavor in the documentation.

**Comment by apparentlymart** [comment]
@copumpkin no, it is dependent on the capabilities of the underlying API. For example, AWS tags on many resource are an arbitrary map from strings to strings, so you can store additional information there that can be used by outside automation.

**Comment by simpsora** [comment]
I'd also really like to have this functionality.  Here's my use case:

I use the Cloudflare provider to manage DNS records.  I'd like to have certain metadata about each record stored with that record (ownership within the organization, purpose, etc.), which is not something the current provider supports.  I'd also like to extend that concept to other resources which don't support any sort of additional metadata either.

I'm ambivalent about whether the metadata is stored in the state file.

**Comment by mildred** [comment]
This could be implemented by a resource. I believe https://github.com/terraform-providers/terraform-provider-null/pull/15 can solve this problem.

Just store your current deployment color in the null-resource, and invert the value each time. Something like:

```hcl
resource "null_resource" "deployment_color" {
  triggers {
    always = "${timestamp()}"
  }
  inputs {
    color = "${null_resource.deployment_color.values.color == "blue" ? "green" : "blue"}"
  }
}
```

It uses a self referencing resource though, don't know if it works.

But such mechanisms could be great for a number of use cases. On top of my head, I have:

- specify a value that would be computed the first time terraform is appleid, then would keep always the same (could be useful to compute the `-bootstrap-expect=` CLI flag for consul servers of a variable sized Consul cluster)
- Implement blue/green for certificates of a private authority
